### ISSUE TITLE
Workflow stuck in `Running` when Karpenter `pod deleted`

### ISSUE THREAD

**Issue by panicboat** [argoproj/argo-workflows#13152]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

We are using Argo Workflows in an EKS + Karpenter environment and we occasionally get `pod deleted` in this environment.
This is caused by Karpenter deleting pods when it tries to replace a node.
If a pod is deleted outside the control of Argo Workflows in this manner, the workflow does not terminate and remains  `Running`.
```console
% argo get update-assignments-1715907600 -n develop-workflow
Name:                update-assignments-1715907600
Namespace:           develop-workflow
ServiceAccount:      orphan-jobs-coaching
Status:              Running
Conditions:          
 PodRunning          False
Created:             Fri May 17 10:00:00 +0900 (3 weeks ago)
Started:             Fri May 17 10:00:01 +0900 (3 weeks ago)
Duration:            3 weeks 2 hours
EstimatedDuration:   2 minutes 15 seconds
Progress:            1/2
ResourcesDuration:   5s*(100Mi memory),0s*(1 cpu)

STEP                                        TEMPLATE              PODNAME                                                        DURATION  MESSAGE
 ⚠ update-assignments-1715907600(0)         main                  update-assignments-1715907600-main-2683904787                  6m        pod deleted                            
                                                                                                                                                                                           
 ✔ update-assignments-1715907600.onExit(0)  exit-handler                                                                                                                          
 └─┬─✔ datadog-metrics(0)                   argo-workflows-hooks  update-assignments-1715907600-argo-workflows-hooks-1115121203  41s                                              
   ├─○ deadmans-snitch                      argo-workflows-hooks                                                                           when ''none' != none' evaluated false  
   └─○ slack-notification                   argo-workflows-hooks                                                                           when 'none == "Always" ||
(none == "Failure" && Error != "Succeeded")
' evaluated false
```
If a CronWorkflow with concurrencyPolicy of Forbid is in this state, the Workflow may not be started the next time it is scheduled.

### Version
v3.5.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: CronWorkflow
metadata:
  name: update-assignments
spec:
  suspend: false
  schedule: "*/10 0,7-23 * * *"
  concurrencyPolicy: Forbid
  workflowSpec:
    entrypoint: main
    templates:
      - name: main
        container:
          name: main
          image: docker/whalesay
          command: [cowsay]
          args: ['Hello, World!']
````

<details>
<summary>Actual manifests</summary>

```yaml
metadata:
  name: update-assignments-1715907600
  namespace: develop-workflow
  uid: ce00fd8e-f479-43af-a807-0bbb8fab3a68
  resourceVersion: '861225988'
  generation: 11
  creationTimestamp: '2024-05-17T01:00:00Z'
  labels:
    workflows.argoproj.io/completed: 'false'
    workflows.argoproj.io/cron-workflow: update-assignments
    workflows.argoproj.io/name: update-assignments
    workflows.argoproj.io/owner: coaching-web-devs
    workflows.argoproj.io/phase: Running
    workflows.argoproj.io/project: none
    workflows.argoproj.io/service: orphan-jobs
  annotations:
    workflows.argoproj.io/pod-name-format: v2
    workflows.argoproj.io/scheduled-time: '2024-05-17T01:00:00Z'
    workflows.argoproj.io/slack-notification-on: none
    workflows.argoproj.io/snitch-token: none
  ownerReferences:
    - apiVersion: argoproj.io/v1alpha1
      kind: CronWorkflow
      name: update-assignments
      uid: dcf50917-aec5-487d-9dce-513f7f7d0141
      controller: true
      blockOwnerDeletion: true
  managedFields:
    - manager: workflow-controller
      operation: Update
      apiVersion: argoproj.io/v1alpha1
      time: '2024-05-24T01:10:01Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:workflows.argoproj.io/pod-name-format: {}
            f:workflows.argoproj.io/scheduled-time: {}
            f:workflows.argoproj.io/slack-notification-on: {}
            f:workflows.argoproj.io/snitch-token: {}
          f:labels:
            .: {}
            f:workflows.argoproj.io/completed: {}
            f:workflows.argoproj.io/cron-workflow: {}
            f:workflows.argoproj.io/name: {}
            f:workflows.argoproj.io/owner: {}
            f:workflows.argoproj.io/phase: {}
            f:workflows.argoproj.io/project: {}
            f:workflows.argoproj.io/service: {}
          f:ownerReferences:
            .: {}
            k:{"uid":"dcf50917-aec5-487d-9dce-513f7f7d0141"}: {}
        f:spec: {}
        f:status: {}
spec:
  templates:
    - name: main
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: main
        image: >-
          575405593970.dkr.ecr.ap-northeast-1.amazonaws.com/monorepo/orphan-jobs@sha256:cc2244fb942815c9b2b9b85ba07843c53fde999db31dc854b18d4e5a641b4268
        command:
          - bundle
          - exec
          - rake
        args:
          - batches:coaching:update_assignments
        envFrom:
          - configMapRef:
              name: orphan-jobs-config
          - configMapRef:
              name: orphan-jobs-coaching-config
          - secretRef:
              name: orphan-jobs-498a7bf4-ab44-4d2d-a1f4-47df298bb48b
        env:
          - name: DRY_RUN
            value: 'false'
          - name: DD_DOGSTATSD_SOCKET
            value: /var/run/datadog/dogstatsd.socket
          - name: AWS_STS_REGIONAL_ENDPOINTS
            value: regional
          - name: SHIFT_GRADE
            value: 'false'
          - name: CHECK_INITIAL_DAYS
            value: 'false'
          - name: DISABLE_DUPLICATE_COACHING_ASSIGNMENT_UPDATES
            value: 'true'
          - name: COACHING_MASCOT_BOT_EMAIL
            value: ''
          - name: MASCOT_BOT_EMAIL
            value: ''
        resources:
          limits:
            memory: 4000Mi
          requests:
            cpu: 10m
            memory: 2000Mi
        volumeMounts:
          - name: dsdsocket
            readOnly: true
            mountPath: /var/run/datadog
    - name: exit-handler
      inputs: {}
      outputs: {}
      metadata: {}
      steps:
        - - name: datadog-metrics
            template: argo-workflows-hooks
            arguments:
              parameters:
                - name: subcommand
                  value: datadog-metrics
          - name: slack-notification
            template: argo-workflows-hooks
            arguments:
              parameters:
                - name: subcommand
                  value: slack-notification
            when: >
              {{
              workflow.annotations.workflows.argoproj.io/slack-notification-on
              }} == "Always" ||

              ({{
              workflow.annotations.workflows.argoproj.io/slack-notification-on
              }} == "Failure" && {{ workflow.status }} != "Succeeded")
          - name: deadmans-snitch
            template: argo-workflows-hooks
            arguments:
              parameters:
                - name: subcommand
                  value: snitch-checkin
            when: >-
              '{{ workflow.annotations.workflows.argoproj.io/snitch-token }}' !=
              none
    - name: argo-workflows-hooks
      inputs:
        parameters:
          - name: subcommand
          - name: base-url
            default: https://workflows.eks.stg.sapuri.dev
          - name: extra-tags
            default: aws_account:sapuri-staging,environment:staging
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: >-
          231380575420.dkr.ecr.ap-northeast-1.amazonaws.com/system-components/argo-workflows-hooks:main
        args:
          - '{{ inputs.parameters.subcommand }}'
        env:
          - name: EXTRA_TAGS
            value: '{{ inputs.parameters.extra-tags }}'
          - name: WORKFLOW_BASE_URL
            value: '{{ inputs.parameters.base-url }}'
          - name: WORKFLOW_NAME
            value: '{{ workflow.name }}'
          - name: WORKFLOW_NAMESPACE
            value: '{{ workflow.namespace }}'
          - name: WORKFLOW_UID
            value: '{{ workflow.uid }}'
          - name: WORKFLOW_ANNOTATIONS_JSON
            value: '{{ workflow.annotations.json }}'
          - name: WORKFLOW_LABELS_JSON
            value: '{{ workflow.labels.json }}'
          - name: WORKFLOW_CREATION_TIMESTAMP
            value: '{{ workflow.creationTimestamp }}'
          - name: WORKFLOW_DURATION
            value: '{{ workflow.duration }}'
          - name: WORKFLOW_STATUS
            value: '{{ workflow.status }}'
          - name: WORKFLOW_FAILURES_JSON
            value: '{{ workflow.failures }}'
          - name: SLACK_BOT_TOKEN
            valueFrom:
              secretKeyRef:
                name: argo-workflows-secret
                key: SLACK_BOT_TOKEN
          - name: DD_API_KEY
            valueFrom:
              secretKeyRef:
                name: argo-workflows-secret
                key: DD_API_KEY
        resources: {}
        imagePullPolicy: Always
      retryStrategy:
        retryPolicy: Always
  entrypoint: main
  arguments: {}
  serviceAccountName: orphan-jobs-coaching
  volumes:
    - name: dsdsocket
      hostPath:
        path: /var/run/datadog/
  nodeSelector:
    karpenter.sh/capacity-type: on-demand
  onExit: exit-handler
  ttlStrategy:
    secondsAfterCompletion: 300
    secondsAfterSuccess: 300
    secondsAfterFailure: 300
  podGC:
    strategy: OnWorkflowCompletion
  securityContext:
    fsGroup: 1000
  metrics:
    prometheus:
      - name: exec_duration_gauge
        labels:
          - key: owner
            value: coaching-web-devs
          - key: service
            value: orphan-jobs
          - key: project
            value: none
          - key: workflow_namespace
            value: develop-workflow
          - key: workflow_name
            value: update-assignments
          - key: workflow_origin_name
            value: update-assignments-1715907600
          - key: workflow_uid
            value: ce00fd8e-f479-43af-a807-0bbb8fab3a68
          - key: status
            value: '{{ workflow.status }}'
        help: Duration of execution by name
        gauge:
          value: '{{ workflow.duration }}'
          realtime: null
      - name: exec_status_count
        labels:
          - key: owner
            value: coaching-web-devs
          - key: service
            value: orphan-jobs
          - key: project
            value: none
          - key: workflow_namespace
            value: develop-workflow
          - key: workflow_name
            value: update-assignments
          - key: workflow_origin_name
            value: update-assignments-1715907600
          - key: workflow_uid
            value: ce00fd8e-f479-43af-a807-0bbb8fab3a68
          - key: status
            value: '{{ workflow.status }}'
        help: Count of execution by status
        counter:
          value: '1'
  shutdown: Terminate
  templateDefaults:
    inputs: {}
    outputs: {}
    metadata:
      annotations:
        karpenter.sh/do-not-disrupt: 'true'
      labels:
        workflows.argoproj.io/name: update-assignments
        workflows.argoproj.io/owner: coaching-web-devs
        workflows.argoproj.io/project: none
        workflows.argoproj.io/service: orphan-jobs
    retryStrategy:
      limit: '10'
      retryPolicy: OnTransientError
      backoff:
        duration: '1'
        factor: '2'
        maxDuration: 10m
    metrics:
      prometheus:
        - name: template_duration_gauge
          labels:
            - key: owner
              value: coaching-web-devs
            - key: service
              value: orphan-jobs
            - key: project
              value: none
            - key: workflow_namespace
              value: develop-workflow
            - key: workflow_name
              value: update-assignments
            - key: workflow_origin_name
              value: update-assignments-1715907600
            - key: workflow_uid
              value: ce00fd8e-f479-43af-a807-0bbb8fab3a68
            - key: status
              value: '{{ status }}'
          help: Duration of execution by name
          gauge:
            value: '{{ workflow.duration }}'
            realtime: null
        - name: template_status_count
          labels:
            - key: owner
              value: coaching-web-devs
            - key: service
              value: orphan-jobs
            - key: project
              value: none
            - key: workflow_namespace
              value: develop-workflow
            - key: workflow_name
              value: update-assignments
            - key: workflow_origin_name
              value: update-assignments-1715907600
            - key: workflow_uid
              value: ce00fd8e-f479-43af-a807-0bbb8fab3a68
            - key: status
              value: '{{ status }}'
          help: Count of step execution by result status
          counter:
            value: '1'
  artifactGC:
    strategy: Never
status:
  phase: Running
  startedAt: '2024-05-17T01:00:01Z'
  finishedAt: null
  estimatedDuration: 135
  progress: 1/2
  nodes:
    update-assignments-1715907600:
      id: update-assignments-1715907600
      name: update-assignments-1715907600
      displayName: update-assignments-1715907600
      type: Retry
      templateName: main
      templateScope: local/update-assignments-1715907600
      phase: Error
      message: pod deleted
      startedAt: '2024-05-17T01:00:01Z'
      finishedAt: '2024-05-17T01:06:32Z'
      progress: 0/1
      children:
        - update-assignments-1715907600-2683904787
    update-assignments-1715907600-1115121203:
      id: update-assignments-1715907600-1115121203
      name: update-assignments-1715907600.onExit(0)[0].datadog-metrics(0)
      displayName: datadog-metrics(0)
      type: Pod
      templateName: argo-workflows-hooks
      templateScope: local/update-assignments-1715907600
      phase: Succeeded
      boundaryID: update-assignments-1715907600-1380424462
      startedAt: '2024-05-17T01:06:33Z'
      finishedAt: '2024-05-17T01:07:14Z'
      progress: 1/1
      resourcesDuration:
        cpu: 0
        memory: 5
      nodeFlag:
        retried: true
      inputs:
        parameters:
          - name: subcommand
            value: datadog-metrics
          - name: base-url
            default: https://workflows.eks.stg.sapuri.dev
            value: https://workflows.eks.stg.sapuri.dev
          - name: extra-tags
            default: aws_account:sapuri-staging,environment:staging
            value: aws_account:sapuri-staging,environment:staging
      outputs:
        artifacts:
          - name: main-logs
            s3:
              key: >-
                update-assignments-1715907600/update-assignments-1715907600-argo-workflows-hooks-1115121203/main.log
        exitCode: '0'
      hostNodeName: ip-10-34-184-138.ap-northeast-1.compute.internal
    update-assignments-1715907600-1380424462:
      id: update-assignments-1715907600-1380424462
      name: update-assignments-1715907600.onExit(0)
      displayName: update-assignments-1715907600.onExit(0)
      type: Steps
      templateName: exit-handler
      templateScope: local/update-assignments-1715907600
      phase: Succeeded
      startedAt: '2024-05-17T01:06:33Z'
      finishedAt: '2024-05-17T01:07:20Z'
      progress: 1/1
      resourcesDuration:
        cpu: 0
        memory: 5
      nodeFlag:
        hooked: true
        retried: true
      children:
        - update-assignments-1715907600-879461520
      outboundNodes:
        - update-assignments-1715907600-1115121203
        - update-assignments-1715907600-3780031854
        - update-assignments-1715907600-3502201003
    update-assignments-1715907600-2373354307:
      id: update-assignments-1715907600-2373354307
      name: update-assignments-1715907600.onExit
      displayName: update-assignments-1715907600.onExit
      type: Retry
      templateName: exit-handler
      templateScope: local/update-assignments-1715907600
      phase: Succeeded
      startedAt: '2024-05-17T01:06:33Z'
      finishedAt: '2024-05-17T01:07:20Z'
      progress: 1/1
      resourcesDuration:
        cpu: 0
        memory: 5
      nodeFlag:
        hooked: true
        retried: true
      children:
        - update-assignments-1715907600-1380424462
    update-assignments-1715907600-2545805072:
      id: update-assignments-1715907600-2545805072
      name: update-assignments-1715907600.onExit(0)[0].datadog-metrics
      displayName: datadog-metrics
      type: Retry
      templateName: argo-workflows-hooks
      templateScope: local/update-assignments-1715907600
      phase: Succeeded
      boundaryID: update-assignments-1715907600-1380424462
      startedAt: '2024-05-17T01:06:33Z'
      finishedAt: '2024-05-17T01:07:20Z'
      progress: 1/1
      resourcesDuration:
        cpu: 0
        memory: 5
      inputs:
        parameters:
          - name: subcommand
            value: datadog-metrics
          - name: base-url
            default: https://workflows.eks.stg.sapuri.dev
            value: https://workflows.eks.stg.sapuri.dev
          - name: extra-tags
            default: aws_account:sapuri-staging,environment:staging
            value: aws_account:sapuri-staging,environment:staging
      outputs:
        artifacts:
          - name: main-logs
            s3:
              key: >-
                update-assignments-1715907600/update-assignments-1715907600-argo-workflows-hooks-1115121203/main.log
        exitCode: '0'
      children:
        - update-assignments-1715907600-1115121203
    update-assignments-1715907600-2683904787:
      id: update-assignments-1715907600-2683904787
      name: update-assignments-1715907600(0)
      displayName: update-assignments-1715907600(0)
      type: Pod
      templateName: main
      templateScope: local/update-assignments-1715907600
      phase: Error
      message: pod deleted
      startedAt: '2024-05-17T01:00:01Z'
      finishedAt: '2024-05-17T01:06:31Z'
      progress: 0/1
      nodeFlag:
        retried: true
      hostNodeName: ip-10-34-91-8.ap-northeast-1.compute.internal
    update-assignments-1715907600-3502201003:
      id: update-assignments-1715907600-3502201003
      name: update-assignments-1715907600.onExit(0)[0].deadmans-snitch
      displayName: deadmans-snitch
      type: Skipped
      templateName: argo-workflows-hooks
      templateScope: local/update-assignments-1715907600
      phase: Skipped
      boundaryID: update-assignments-1715907600-1380424462
      message: when ''none' != none' evaluated false
      startedAt: '2024-05-17T01:06:39Z'
      finishedAt: '2024-05-17T01:06:39Z'
      nodeFlag: {}
    update-assignments-1715907600-3780031854:
      id: update-assignments-1715907600-3780031854
      name: update-assignments-1715907600.onExit(0)[0].slack-notification
      displayName: slack-notification
      type: Skipped
      templateName: argo-workflows-hooks
      templateScope: local/update-assignments-1715907600
      phase: Skipped
      boundaryID: update-assignments-1715907600-1380424462
      message: |-
        when 'none == "Always" ||
        (none == "Failure" && Error != "Succeeded")
        ' evaluated false
      startedAt: '2024-05-17T01:06:39Z'
      finishedAt: '2024-05-17T01:06:39Z'
      nodeFlag: {}
    update-assignments-1715907600-879461520:
      id: update-assignments-1715907600-879461520
      name: update-assignments-1715907600.onExit(0)[0]
      displayName: '[0]'
      type: StepGroup
      templateScope: local/update-assignments-1715907600
      phase: Succeeded
      boundaryID: update-assignments-1715907600-1380424462
      startedAt: '2024-05-17T01:06:33Z'
      finishedAt: '2024-05-17T01:07:20Z'
      progress: 1/1
      resourcesDuration:
        cpu: 0
        memory: 5
      nodeFlag: {}
      children:
        - update-assignments-1715907600-2545805072
        - update-assignments-1715907600-3780031854
        - update-assignments-1715907600-3502201003
  conditions:
    - type: PodRunning
      status: 'False'
  resourcesDuration:
    cpu: 0
    memory: 5
  artifactRepositoryRef:
    configMap: artifact-repositories
    key: default-s3-artifact-repository
    namespace: develop-workflow
    artifactRepository:
      archiveLogs: true
      s3:
        endpoint: s3.amazonaws.com
        bucket: sapuri-staging-workflow-artifacts-develop
        region: ap-northeast-1
        insecure: false
  artifactGCStatus:
    notSpecified: true
  taskResultsCompletionStatus:
    update-assignments-1715907600-1115121203: true
    update-assignments-1715907600-2683904787: false
```
</details>

### Logs from the workflow controller

```text
N/A
```


### Logs from in your workflow's wait container

```text
N/A
```

**Comment by agilgur5** [comment]
This might be due to the same root cause as #12993

**Comment by panicboat** [comment]
Thanks for checking.
Do you think I can close this as a duplicate?

**Comment by agilgur5** [comment]
Can you verify it has the same symptoms: that you have an incomplete `WorkflowTaskResult`?

**Comment by panicboat** [comment]
I could not get it, so let me leave this Issue as Open.
```sh
% k get workflowtaskresults -n develop-workflow
No resources found in develop-workflow namespace.
% k get workflowtaskresult -n develop-workflow 
No resources found in develop-workflow namespace.
```

**Comment by agilgur5** [comment]
Oh it turns out this is a duplicate of a much older issue: #8783. That was fixed/completed in #12413, which will be released in 3.6.0
